src/ucs/datastruct/ptr_array.c

```diff
@@ -200,9 +199,50 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value)
     return element_index;
 }
 
+void ucs_ptr_array_set(ucs_ptr_array_t *ptr_array, unsigned element_index,
+                       void *new_val)
+{
+    ucs_ptr_array_elem_t *elem;
+    unsigned next, free_iter, free_ahead;
+
+    if (ucs_unlikely(element_index > ptr_array->size)) {
+        ucs_ptr_array_grow(ptr_array, element_index + 1
+                           UCS_MEMTRACK_NAME(ptr_array->name));
+    } else if (!__ucs_ptr_array_is_free(ptr_array->start[element_index])) {
+        ptr_array->start[element_index] = (uintptr_t)new_val;
+        return;
+    }
+
+    next = ucs_ptr_array_freelist_get_next(ptr_array->start[element_index]);
+    ptr_array->start[element_index] = (uintptr_t)new_val;
+
+    /* update the "next index" in the free list */
+    free_iter = ptr_array->freelist;
+    if (ucs_unlikely(free_iter == element_index)) {
```

Comment:
why ucs_unlikely?

Reply:
I find it unlikely that the call to set() happens to pass the last index removed by remove(). It makes no sense to call `remove(i); set(i)` - because you might as well call `set(i)` and overwrite the cell without calling `remove(i)` first.